cmd/lotus-chainwatch/storage.go

```diff
@@ -660,94 +644,123 @@ func (st *storage) storeMinerSectorsHeads(minerTips map[types.TipSetKey][]*miner
+				if err != nil {
+					log.Debugw("error getting miner sector diff", "miner", miner.addr, "error", err)
+					continue
+				}
+				if !changed {
+					continue
+				}
+				changes := val.(*state.MinerSectorChanges)
+				log.Infow("sector changes for miner", "miner", miner.addr.String(), "Added", len(changes.Added), "Extended", len(changes.Extended), "Removed", len(changes.Removed), "oldState", miner.parentTsKey, "newState", miner.tsKey)
+
+				for _, extended := range changes.Extended {
+					if _, err := updateStmt.Exec(nil, extended.To.Info.Expiration, miner.addr.String(), extended.To.Info.SectorNumber); err != nil {
+						return err
+					}
+					if _, err := eventStmt.Exec(miner.addr.String(), extended.To.Info.SectorNumber, miner.stateroot.String(), "EXTENDED"); err != nil {
+						return err
+					}
+					log.Info("sector extended", "miner", miner.addr.String(), "sector", extended.To.Info.SectorNumber, "old", extended.To.Info.Expiration, "new", extended.From.Info.Expiration)
+				}
+				curTs, err := api.ChainGetTipSet(context.TODO(), miner.tsKey)
+				if err != nil {
+					return err
+				}
+
+				for _, removed := range changes.Removed {
+					// decide if they were terminated or extended
+					if removed.Info.Expiration > curTs.Height() {
+						if _, err := updateStmt.Exec(nil, removed.Info.Expiration, miner.addr.String(), removed.Info.SectorNumber); err != nil {
+							return err
+						}
+						if _, err := eventStmt.Exec(miner.addr.String(), removed.Info.SectorNumber, miner.stateroot.String(), "TERMINATED"); err != nil {
+							return err
+						}
+						log.Info("sector terminated", "miner", miner.addr.String(), "sector", removed.Info.SectorNumber, "old", "sectorExpiration", removed.Info.Expiration, "terminationEpoch", curTs.Height())
+					}
+					if _, err := eventStmt.Exec(miner.addr.String(), removed.Info.SectorNumber, miner.stateroot.String(), "EXPIRED"); err != nil {
+						return err
+					}
+					log.Info("sector removed", "miner", miner.addr.String(), "sector", removed.Info.SectorNumber, "old", "sectorExpiration", removed.Info.Expiration, "currEpoch", curTs.Height())
+				}
+
+				for _, added := range changes.Added {
+					if _, err := eventStmt.Exec(miner.addr.String(), added.Info.SectorNumber, miner.stateroot.String(), "ADDED"); err != nil {
+						return err
+					}
+				}
 			}
-			return nil
-		}); err != nil {
-			log.Warnw("old sectors foreach", "error", err)
-			return err
-		}
-		if len(deletedSectors) > 0 {
-			log.Infow("Calculated updates", "miner", miner.minerState.addr, "deleted sectors", len(deletedSectors))
 		}
 	}
-	// now we have all the sectors that were removed, update the database
-	tx, err := st.db.Begin()
-	if err != nil {
+
+	if err := updateStmt.Close(); err != nil {
 		return err
 	}
-	stmt, err := tx.Prepare(`UPDATE miner_sectors SET termination_epoch=$1 WHERE miner_id=$2 AND sector_id=$3`)
-	if err != nil {
+
+	if err := eventStmt.Close(); err != nil {
```

Comment:
Does this work? I don't think you can have two prepared statements nested within each other?

Reply:
Postgres says you can, but suggests using savepoints: https://www.postgresql.org/docs/9.4/sql-begin.html
> Issuing BEGIN when already inside a transaction block will provoke a warning message. The state of the transaction is not affected. To nest transactions within a transaction block, use savepoints (see SAVEPOINT).

I'll refactor this to perform the update in a separate loop.